onnxruntime/core/framework/sparse_tensor.cc

```diff
@@ -3,31 +3,441 @@
+  auto mutator = MakeBlockSparseData(values_shape, index_shape);
+  if (values_shape.Size() > 0) {
+    auto& dst_values = mutator.Values();
+    auto& dst_index = mutator.Index();
+    Tensor src_values(dst_values.DataType(), dst_values.Shape(), const_cast<void*>(values_data), data_location);
+    Tensor src_index(dst_index.DataType(), dst_index.Shape(), const_cast<int32_t*>(index_data), data_location);
+    ORT_RETURN_IF_ERROR(CopyData(data_transfer, MakeListConst(src_values, src_index), MakeListNonConst(dst_values, dst_index)));
+  }
+  return Status::OK();
+}
+
+SparseTensor::BlockSparseMutator SparseTensor::MakeBlockSparseData(const TensorShape& values_shape, const TensorShape& index_shape) {
+  ORT_ENFORCE(allocator_ != nullptr, "This method should follow a call to constructor that supplies the allocator");
+  ORT_THROW_IF_ERROR(ValidateBlockSparseShapes(values_shape, index_shape));
+  if (values_shape.Size() > 0) {
+    const auto data_size = values_shape.Size() * ml_data_type_->Size();
+    const auto index_size = index_shape.Size() * sizeof(int32_t);
+    const auto required_buffer_size = CalculateRequiredBufferSize(gsl::narrow<int64_t>(data_size),
+                                                                  gsl::narrow<int64_t>(index_size));
+    ORT_THROW_IF_ERROR(AllocateBuffer(required_buffer_size, static_cast<size_t>(data_size / ml_data_type_->Size())));
+  }
+  values_ = Tensor(DataType(), values_shape, p_data_, Location());
+  format_data_.resize(1);
+  format_data_[0] = Tensor(DataTypeImpl::GetType<int32_t>(), index_shape, IndicesStart(values_.SizeInBytes()), Location());
+  format_ = SparseFormat::kBlockSparse;
+  return BlockSparseMutator(values_, format_data_[0]);
+}
+
+Status SparseTensor::Copy(const DataTransferManager& data_transfer_manager, int exec_q_id, SparseTensor& dst_tensor) const {
+  const IDataTransfer* data_transfer = data_transfer_manager.GetDataTransfer(Location().device,
+                                                                             dst_tensor.Location().device);
+  ORT_RETURN_IF_NOT(data_transfer != nullptr, "Unable to find a data transfer for copying from device type: ",
+                    Location().device.Type(), " to device type: ", dst_tensor.Location().device.Type());
+
+  return Copy(*data_transfer, dst_tensor, exec_q_id);
+}
+
+Status SparseTensor::Copy(const IDataTransfer& data_transfer, SparseTensor& dst_tensor, int exec_q_id) const {
+  // Do not copy same destination
+  if (this == &dst_tensor) {
+    return Status::OK();
+  }
+
+  ORT_RETURN_IF_NOT(format_ != SparseFormat::kUndefined, "This instance should not be empty");
+  const bool is_string = IsDataTypeString();
+
+  ORT_RETURN_IF_NOT(dst_tensor.Format() == SparseFormat::kUndefined, "Destination should be empty");
+  ORT_RETURN_IF_NOT(dst_tensor.allocator_ != nullptr, "Destination must have a CPU allocator set");
+  ORT_RETURN_IF_NOT((!is_string || dst_tensor.Location().device.Type() == OrtDevice::CPU),
```

Comment:
Does it have to be a CPU allocator or just an allocator?

Reply:
Strings can only be copied to a memory allocated on CPU. So it is any allocator that is on CPU.